ruyi/cli/completion.py

```diff
@@ -0,0 +1,74 @@
+"""
+helper functions for CLI completions
+"""
+
+from typing import cast, Callable, Optional, Protocol, Sequence, Any, TYPE_CHECKING
+import argparse
+
+from argcomplete.completers import BaseCompleter
+
+if TYPE_CHECKING:
+    from .. import config
+
+
+class ArgcompleteAction(argparse.Action):
+    # see https://github.com/kislyuk/argcomplete/issues/443 for why this typing is needed
+    completer: Optional[Callable[[str, object], list[str]]]
+
+    def __call__(
+        self,
+        parser: argparse.ArgumentParser,
+        namespace: argparse.Namespace,
+        values: str | Sequence[Any] | None,
+        option_string: str | None = None,
+    ) -> None:
+        raise NotImplementedError('.__call__() not defined')
+
+
+class SelfArgumentParser(argparse.ArgumentParser):
```

Comment:
What does the "self" part mean? To me it seems only intended for `ruyi self` actions which is definitely not true. If this is intended to be the application-level parser to use, maybe just `ArgumentParser` is enough -- the imports will show the originating package is not `argparse` but relative from the project root.